### ISSUE TITLE
Special case parameter value 0 scale

### ISSUE THREAD

**Issue by cdeil** [gammapy/gammapy#1698]
I tried to fit this model:
```
spatial_model = SkyPointSource(
    lon_0='0 deg',
    lat_0='0 deg',
)
```
and got this error:
```
/Users/deil/work/code/gammapy/gammapy/utils/modeling.py:391: RuntimeWarning: divide by zero encountered in log10
  scale = 10 ** int(np.log10(value))
---------------------------------------------------------------------------
OverflowError                             Traceback (most recent call last)
<timed exec> in <module>()

~/work/code/gammapy/gammapy/cube/fit.py in fit(self, opts_minuit)
     99         minuit = fit_iminuit(parameters=self.model.parameters,
    100                              function=self.total_stat,
--> 101                              opts_minuit=opts_minuit)
    102         self._minuit = minuit
    103 

~/work/code/gammapy/gammapy/utils/fitting/iminuit.py in fit_iminuit(parameters, function, opts_minuit)
     35 
     36     if parameters.covariance is None:
---> 37         parameters.scale()
     38 
     39     minuit_func = MinuitFunction(function, parameters)

~/work/code/gammapy/gammapy/utils/modeling.py in scale(self, method)
    389             if method == 'scale10':
    390                 value = par.value
--> 391                 scale = 10 ** int(np.log10(value))
    392                 par.factor = value / scale
    393                 par.scale = scale

OverflowError: cannot convert float infinity to integer
```
It happens because both methods that we have to scale parameters cannot handle exact float value of zero.

We've talked about this before, and possible solutions are:
1. Special-case `value == 0` with an exact float assert in an `if` statement and don't scale those parameters
2. Introduce parameter "type" attribute, among which we have "spatial lon" and "spatial lat" and don't scale the parameters of that type.

Obviously 1. is quicker, but if we agree that 2. is a direction in our modeling worth trying out for the coming months, it's also not a lot of work to add.

@joleroi @adonath - Thoughts?

**Comment by adonath** [comment]
Why not just:

```python
try:
    scale = 10 ** int(np.log10(value))
except OverflowError:
    scale = 1.
```
Technically I do not see any issues with this solution...

**Comment by joleroi** [comment]
I still don't really see the purpose of 2., so I'd go for 1. but using ``np.isclose`` not a ``try`` - ``except`` statemant, but that's of course a matter of taste ...

**Comment by adonath** [comment]
I agree with @joleroi, that I do not see the need for solution 2. either.

**Comment by cdeil** [comment]
I think `isclose` could be problematic. Note that we had cases of wanting to fit spectral models with a norm of `1e-20`, which is very close to 0.

I was just thinking that there might be cases where one wants to fit a Gammapy model with a negative parameter. E.g. fitting the Galactic center position could result in  `lon = -0.01 deg` or `lon=+0.01 deg`, and switching to `lon = 359.99 deg` would make the fit unstable with the 0 to 360 jump.

We could also put the hack to leave the lon and lat parameter alone?
```
if 'lon' in par.name or 'lat' in par.name:
    continue
```

I'm still not sure what to put as short-term solution: @joleroi @adonath - please comment again.

---

The main purpose of solution 2, to have parameter types like "spatial" and "spectral" would be to make the MapFit and MapEvaluator smart, so that they cache some things and only re-compute things as needed. E.g. if the optimiser takes a step in a spectral parameter, spatial models and PSF convolution doesn't have to be re-computed. Such a smart model evaluator could give an order of magnitude speedup at the cost of some RAM. I'm not saying we should do that in Gammapy, I just thought it's a direction worth exploring, and having parameter type info would allow for that exploration. Solving the parameter auto scale issue would be more of a nice side benefit, but in the same spirit: make our modeling smarter for our domain-specific models and data.

**Comment by adonath** [comment]
@cdeil Indeed the case of negative parameter values should be handled correctly, which is currently not the case...I agree `np.isclose()` is problematic, I think it's still fine to just catch the `OverflowError` for  the `value=0` case. I'd be also fine with not scaling certain parameters, but maybe I'd rather handle this with a flag on the parameter object, as it could apply for other (especially spectral) parameters such as `index`, or `lambda_` or `alpha` and `beta` as well...

**Comment by adonath** [comment]
I agree it's useful to have smart model evaluators in the future, but there are many ways to achieve what you suggest and it's not clear to me that the parameter type property is needed for this. E.g. models could as well just have a `.state` attribute, that specifies whether the model has changed since the last evaluation...

**Comment by joleroi** [comment]
> I agree np.isclose() is problematic

I agree in this case. I argued against ``try`` - ``except`` because I've encounter numerous examples in gammapy where this policy lead to unforseen errors handled in a bad way. But here it might indeed be a solution

> I agree it's useful to have smart model evaluators in the future,

with emphasis on 'in the future'

**Comment by cdeil** [comment]
> maybe I'd rather handle this with a flag on the parameter object

Basically that's what the proposed Parameter `type` is: a multiple-choice string that for now is used as a flag to decide if / how to scale parameters in this method.

Unless you have time to make a PR, we can just briefly talk about this and then put in something tomorrow before cutting the release.

**Comment by adonath** [comment]
@cdeil I see your point, but I think whether the scaling should be applied or not is independent of the fact whether a parameter is of `spectral` or `spatial` type (e.g. `lon` and `lat` as well `index` wouldn't have to be scaled... ). So wee might need to introduce different or additional categories...or maybe just allow `scale=None` if a parameter should not be scaled?

**Comment by joleroi** [comment]
Why don't we go for @adonath solution
> I think it's still fine to just catch the OverflowError
And put something fancy as soon as the need arises

**Comment by joleroi** [comment]
@cdeil I think this is fixed now, but can you nevertheless try your fit again and confirm?

**Comment by cdeil** [comment]
Fixed by @joleroi in #1708
